### PR TITLE
upgrade rust and tokio + set max threads to num of cpus

### PR DESCRIPTION
**Pull request by depombo**

### CODE REVIEW

**Comment by dfellis** [comment]
This doesn't actually fix the real issue of #354 which is that we're still creating a tokio task per array element, which explodes the AVM with an OOM on large arrays unnecessarily (also because we duplicate the working memory for each task right now, so we get array size to the array size amount of memory consumption)

File: avm/Cargo.toml [hunk 1]
```diff
@@ -5,7 +5,7 @@ license = "AGPL-3.0"
 homepage = "https://alan-lang.org"
 documentation = "https://docs.alan-lang.org"
 repository = "https://github.com/alantech/alan"
-version = "0.1.22"
+version = "0.1.48"
```
Comment (dfellis) [inline avm/Cargo.toml:8]: Is this the Rust version? Or why did you change it?

File: avm/Cargo.toml [hunk 2]
```diff
@@ -5,7 +5,7 @@ license = "AGPL-3.0"
...
 authors = ["Luis de Pombo <luis@alantechnologies.com>", "David Ellis <david@alantechnologies.com>"]
 edition = "2018"
```
Comment (dfellis) [inline avm/Cargo.toml:10]: I assumed *this* was the field you need to change?

File: avm/Cargo.toml (hunk 1 above)
Comment (depombo) [reply avm/Cargo.toml:8]: this is the Rust version

File: avm/Cargo.toml (hunk 2 above)
Comment (depombo) [reply avm/Cargo.toml:10]: Initially, I thought so too

File: avm/Cargo.toml (hunk 1 above)
Comment (dfellis) [reply avm/Cargo.toml:8]: Not according to: https://doc.rust-lang.org/cargo/reference/manifest.html#the-version-field

Comment (dfellis) [reply avm/Cargo.toml:8]: It appears you *can't* specify the Rust compiler version in Cargo? https://github.com/emk/heroku-buildpack-rust/issues/11 That explains why we've had such difficulties with the submitted projects compiling on our laptops.

Comment (dfellis) [reply avm/Cargo.toml:8]: Tracing through several years of github issues leads me to: https://github.com/rust-lang/rustup/pull/1172

We apparently need to add a `rust-toolchain` file that contains, as far as I can tell, only the version of rust that we want to use?

**Comment by dfellis** [comment]
It looks like this upgrade also broke the HTTP client and server code :/

**Comment by depombo** [comment]
> This doesn't actually fix the real issue of #354 which is that we're still creating a tokio task per array element, which explodes the AVM with an OOM on large arrays unnecessarily (also because we duplicate the working memory for each task right now, so we get array size to the array size amount of memory consumption)

I see. To some extent I did not understand the issue correctly. It should not spin one thread per array element, but also it should not spin on task per array element. Since tokio uses a work stealing mechanism this addresses the former, but not the latter.

**Comment by depombo** [comment]
> It looks like this upgrade also broke the HTTP client and server code :/

Yeah, I think this is a bug between `enable_all` not actually turning on `enable_io`

**Comment by dfellis** [comment]
> > It looks like this upgrade also broke the HTTP client and server code :/
> 
> Yeah, I think this is a bug between `enable_all` not actually turning on `enable_io`

...that's crazy! What else would you mean by "enable_all"?

**Comment by depombo** [comment]
> > > It looks like this upgrade also broke the HTTP client and server code :/
> > 
> > 
> > Yeah, I think this is a bug between `enable_all` not actually turning on `enable_io`
> 
> ...that's crazy! What else would you mean by "enable_all"?

So it turns out that it was silently failing because the latest version of `hyper` and `reqwst` are still using tokio 0.2. `hyper`'s master branch already made the change, but until their next release we can't upgrade. I'm going to abandon and make a new PR where I fix the actual issue of spinning up a tokio task per array element and set max threads without upgrading tokio
